From c55fdd7da20a56cd98e559401c8319df2c602217 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Wed, 28 Oct 2020 17:02:05 +0100
Subject: [PATCH] PONRTSYS-8201: Do not block soft IRQs in locking

As the notifier call is now not using a RCU lock and the
counter update is happening in process context and not soft interrupt
context any more because we changes from timer to workqueue we
do not have to block soft interrupts any more.

With this change the threaded interrupt in the pon mailbox driver is not
getting blocked any more and in my nightly test it was always triggered
in less than 100ms after the hard interrupt.
---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 42 +++++++++++-----------
 1 file changed, 21 insertions(+), 21 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index fbecf0263629..2f4eaebf7474 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -285,16 +285,16 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	 * cumulative counters
 	 */
 	if (initial_update == INITIAL_UPDATE) {
-		write_lock_bh(&cnt->cnt_lock);
+		write_lock(&cnt->cnt_lock);
 		cnt->last.rmon = tmp_counters[0];
 		if (rmon)
 			cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
-		write_unlock_bh(&cnt->cnt_lock);
+		write_unlock(&cnt->cnt_lock);
 		return;
 	}
 
 	/* update 64 bits wide counters based on last read values */
-	write_lock_bh(&cnt->cnt_lock);
+	write_lock(&cnt->cnt_lock);
 #define _DP_ETHTOOL(x, y) cnt->wide_cntrs.y += wraparound(tmp_counters[0].y, cnt->last.rmon.y, sizeof(tmp_counters[0].y));
 	DP_GSW_ETHTOOL_STAT_REG_DECLARE
 #undef _DP_ETHTOOL
@@ -334,7 +334,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	cnt->last.rmon = tmp_counters[0];
 	if (rmon)
 		cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
-	write_unlock_bh(&cnt->cnt_lock);
+	write_unlock(&cnt->cnt_lock);
 
 	devm_kfree(&g_dp_dev->dev, tmp_counters);
 	if (rmon)
@@ -368,12 +368,12 @@ static void update_all_devs_counters(void)
 	/* Fill work_list as a snapshot of the registered_devices_list
 	 * with crucial identifiers only
 	 */
-	read_lock_bh(&dev_list_lock);
+	read_lock(&dev_list_lock);
 	list_for_each(p, &registered_devices_list) {
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
 		wle = kzalloc(sizeof(struct work_list_el), GFP_ATOMIC);
 		if (!wle) {
-			read_unlock_bh(&dev_list_lock);
+			read_unlock(&dev_list_lock);
 			return;
 		}
 		wle->dev = mib_cnt->dev;
@@ -382,7 +382,7 @@ static void update_all_devs_counters(void)
 
 		list_add_tail(&wle->work_list_handle, &work_list);
 	}
-	read_unlock_bh(&dev_list_lock);
+	read_unlock(&dev_list_lock);
 
 	/* Iterate over the work_list, check if the device from work_list still
 	 * exists in registered_devices_list and do counters updated for it.
@@ -390,7 +390,7 @@ static void update_all_devs_counters(void)
 	list_for_each_safe(p, q, &work_list) {
 		wle = list_entry(p, struct work_list_el, work_list_handle);
 		mib_cnt_match = NULL;
-		read_lock_bh(&dev_list_lock);
+		read_lock(&dev_list_lock);
 		/* The dev can not exist anymore in registered_devices_list
 		 * because it could have been removed by dp in meantime.
 		 */
@@ -404,7 +404,7 @@ static void update_all_devs_counters(void)
 				break;
 			}
 		}
-		read_unlock_bh(&dev_list_lock);
+		read_unlock(&dev_list_lock);
 
 		if (mib_cnt_match)
 			DP_DEBUG(DP_DBG_FLAG_MIB,
@@ -461,11 +461,11 @@ s32 add_dev_to_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 
 	INIT_LIST_HEAD(&dev_cnt->dev_list);
 
-	write_lock_bh(&dev_list_lock);
+	write_lock(&dev_list_lock);
 	list_add_tail(&dev_cnt->dev_list, &registered_devices_list);
 	/* read counters for just added dev to save initial value */
 	update_dev_counters(dev_cnt, INITIAL_UPDATE);
-	write_unlock_bh(&dev_list_lock);
+	write_unlock(&dev_list_lock);
 
 	DP_DEBUG(DP_DBG_FLAG_MIB,
 		 "MIB CNT - added device: %s port_id = %d, sub_if_id_grp = %d\n",
@@ -501,7 +501,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 		 "MIB CNT - looking for item to delete: name = %s, port_id = %d, sub_if = %d, inst = %d\n",
 		 dev->name, port_id, sub_if, inst);
 
-	write_lock_bh(&dev_list_lock);
+	write_lock(&dev_list_lock);
 	list_for_each_safe(p, q, &registered_devices_list) {
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
 
@@ -526,7 +526,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 				 mib_cnt->inst);
 		}
 	}
-	write_unlock_bh(&dev_list_lock);
+	write_unlock(&dev_list_lock);
 
 	return 0;
 }
@@ -538,7 +538,7 @@ struct mib_counters *find_registered_dev(s32 port_id, s32 sub_if_id_grp,
 	struct mib_counters *mib_cnt;
 	struct list_head *p;
 
-	read_lock_bh(&dev_list_lock);
+	read_lock(&dev_list_lock);
 	list_for_each(p, &registered_devices_list)
 	{
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
@@ -546,11 +546,11 @@ struct mib_counters *find_registered_dev(s32 port_id, s32 sub_if_id_grp,
 		    (mib_cnt->sub_if_id_grp == sub_if_id_grp) &&
 		    (mib_cnt->inst == inst) &&
 		    (!strcmp(mib_cnt->dev->name, dev->name))) {
-			read_unlock_bh(&dev_list_lock);
+			read_unlock(&dev_list_lock);
 			return mib_cnt;
 		}
 	}
-	read_unlock_bh(&dev_list_lock);
+	read_unlock(&dev_list_lock);
 
 	return NULL;
 }
@@ -644,12 +644,12 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 				port_info->vap_mask);
 
 	/* Find required registered device on list */
-	read_lock_bh(&dev_list_lock);
+	read_lock(&dev_list_lock);
 	mib_cnt = find_registered_dev(ps_subif->port_id, sub_if_id_grp,
 				      ps_subif->inst, dev);
 	if (!mib_cnt) {
 		devm_kfree(&g_dp_dev->dev, ps_subif);
-		read_unlock_bh(&dev_list_lock);
+		read_unlock(&dev_list_lock);
 		return;
 	}
 
@@ -661,7 +661,7 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 	/* Copy wide counters to update by reinserted packets */
 	wide_cnt_mod = mib_cnt->wide_cntrs;
 
-	read_unlock_bh(&dev_list_lock);
+	read_unlock(&dev_list_lock);
 
 	/* Get reinserted packets counters */
 	(void)dp_get_reinsert_cnt_31(mib_cnt->inst,
@@ -782,14 +782,14 @@ static void __exit mib_counters_mod_exit(void)
 	cancel_delayed_work_sync(&cnt_upd_work);
 
 	/* Free all list entries for registered net devices */
-	write_lock_bh(&dev_list_lock);
+	write_lock(&dev_list_lock);
 	list_for_each_safe(p, q, &registered_devices_list)
 	{
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
 		list_del(p);
 		devm_kfree(&g_dp_dev->dev, mib_cnt);
 	}
-	write_unlock_bh(&dev_list_lock);
+	write_unlock(&dev_list_lock);
 
 	/* Deregister ethtool stats cb function */
 	dp_set_ethtool_stats_fn(0, NULL);
